keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,468 @@
+the iptables rules to be stale until they eventually get fixed in a
+future full resync.
+
+In order to detect this, we would have to sanity-check every partial
+restore, to ensure that, if applied to the current kernel state, it
+would generate the same end result as the full restore we would
+otherwise have applied. Although we could do this with the
+`FakeIPTables` code in `pkg/util/iptables/testing`, that code is only
+intended for the sorts of small-ish test cases in the unit tests, and
+not at all optimized for handling the sort of very large rule sets
+that would exist in larger clusters (ie, the rule sets that would
+benefit the most from the partial restore feature).
+
+One possible approach for dealing with that would be to run the
+sanity-checking code only during the Alpha period, and only on
+"smaller" clusters where it could be done without too much performance
+loss. (If the code was going to fail in the way we are checking for
+here, the failure would be triggered by specific types of Services or
+specific types of changes to Services, not by the _number_ of
+Services, so testing only on smaller clusters would still be expected
+to (eventually) detect it.)
+
+We could then add a metric for "partial resync generated a different
+state than full resync", and alert on that.
+
+(We could potentially run this check even on clusters that don't enable
+the partial restores feature: we would generate the partial restore
+for the purpose of checking that it was correct, but then just throw
+it away and use the full restore.)
+
+## Design Details
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+This is the latest in a series of iptables kube-proxy improvements,
+and many prerequisite tests have already been added.
+
+##### Unit tests
+
+We will add unit tests of the partial-resyncing feature, to ensure
+that expected rules are updated when services and endpoints
+change. (This is already done in the current PR.)
+
+Existing coverage:
+- `k8s.io/kubernetes/pkg/proxy/iptables`: `2022-08-02` - `67%`
+
+##### Integration tests
+
+Kube-proxy is mostly tested via unit tests and e2e, not integration,
+and this is not expected to change.
+
+##### e2e tests
+
+All existing kube-proxy/Service e2e tests would be testing the new
+code. In fact, a large percentage of the non-networking e2e tests
+would also end up testing the new code. There is really no good way to
+_specifically_ test this feature in any way that wasn't redundant with
+the existing e2e tests and the feature's unit tests.
+
+It would be useful to test at the end of the e2e test run (or at
+least, at the end of the "Service" test suite) that the "partial sync
+failures" and "partial sync mismatch" metrics are both "0".
+
+### Graduation Criteria
```

Comment:
I think that we should evaluate the impact of the feature, we have metrics already that record the time to sync the rules.
My gut feeling is that for small, medium clusters/#services/endpoints this will not influence much ... and for large clusters we already fixed it with promising results https://github.com/kubernetes/kubernetes/pull/110268#issuecomment-1206214513

Reply:
The improvement seen in that link is from reducing the time spent in `iptables-save`. The partial-restores patch is likely to result in a much larger improvement.

Reply:
but yes, graduation should be dependent on the feature actually being useful :slightly_smiling_face:

Reply:
oh, I see my misunderstanding, I somehow mixed save and restore in my mind